pkg/services/pull_request/bitbucket_server.go

```diff
@@ -0,0 +1,97 @@
+package pull_request
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	bitbucketv1 "github.com/gfleury/go-bitbucket-v1"
+)
+
+type BitbucketService struct {
+	client         *bitbucketv1.APIClient
+	projectKey     string
+	repositorySlug string
+	branchMatch    *regexp.Regexp
+	// Not supported for PRs by Bitbucket Server
+	// labels         []string
+}
+
+var _ PullRequestService = (*BitbucketService)(nil)
+
+func NewBitbucketServiceBasicAuth(ctx context.Context, username, password, url, projectKey, repositorySlug string, branchMatch *string) (PullRequestService, error) {
+	bitbucketConfig := bitbucketv1.NewConfiguration(url)
+	// Avoid the XSRF check
+	bitbucketConfig.AddDefaultHeader("x-atlassian-token", "no-check")
+	bitbucketConfig.AddDefaultHeader("x-requested-with", "XMLHttpRequest")
+
+	ctx = context.WithValue(ctx, bitbucketv1.ContextBasicAuth, bitbucketv1.BasicAuth{
+		UserName: username,
+		Password: password,
+	})
+	return newBitbucketService(ctx, bitbucketConfig, projectKey, repositorySlug, branchMatch)
+}
+
+func NewBitbucketServiceNoAuth(ctx context.Context, url, projectKey, repositorySlug string, branchMatch *string) (PullRequestService, error) {
+	return newBitbucketService(ctx, bitbucketv1.NewConfiguration(url), projectKey, repositorySlug, branchMatch)
+}
+
+func newBitbucketService(ctx context.Context, bitbucketConfig *bitbucketv1.Configuration, projectKey, repositorySlug string, branchMatch *string) (PullRequestService, error) {
+	if !strings.HasSuffix(bitbucketConfig.BasePath, "/rest") {
+		bitbucketConfig.BasePath = bitbucketConfig.BasePath + "/rest"
+	}
+	bitbucketClient := bitbucketv1.NewAPIClient(ctx, bitbucketConfig)
+
+	var branchMatchRegexp *regexp.Regexp
+	if branchMatch != nil {
+		var err error
+		branchMatchRegexp, err = regexp.Compile(*branchMatch)
+		if err != nil {
+			return nil, fmt.Errorf("error compiling BranchMatch regexp %q: %v", *branchMatch, err)
+		}
+	}
+
+	return &BitbucketService{
+		client:         bitbucketClient,
+		projectKey:     projectKey,
+		repositorySlug: repositorySlug,
+		branchMatch:    branchMatchRegexp,
+	}, nil
+}
+
+func (b *BitbucketService) List(_ context.Context) ([]*PullRequest, error) {
+	paged := map[string]interface{}{
+		"limit": 100,
+	}
+
+	pullRequests := []*PullRequest{}
+	for {
+		response, err := b.client.DefaultApi.GetPullRequestsPage(b.projectKey, b.repositorySlug, paged)
+		if err != nil {
+			return nil, fmt.Errorf("error listing pull requests for %s/%s: %v", b.projectKey, b.repositorySlug, err)
+		}
+		pulls, err := bitbucketv1.GetPullRequestsResponse(response)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing pull request response %s: %v", response.Values, err)
```

Comment:
What kind of secrets are you thinking of ? Here's an example of a response from the [docs (`/rest/api/1.0/projects/{projectKey}/repos/{repositorySlug}/pull-requests`)](https://docs.atlassian.com/bitbucket-server/rest/7.17.5/bitbucket-rest.html#idp300):
<details>

```
{
    "size": 1,
    "limit": 25,
    "isLastPage": true,
    "values": [
        {
            "id": 101,
            "version": 1,
            "title": "Talking Nerdy",
            "description": "It’s a kludge, but put the tuple from the database in the cache.",
            "state": "OPEN",
            "open": true,
            "closed": false,
            "createdDate": 1359075920,
            "updatedDate": 1359085920,
            "fromRef": {
                "id": "refs/heads/feature-ABC-123",
                "repository": {
                    "slug": "my-repo",
                    "name": null,
                    "project": {
                        "key": "PRJ"
                    }
                }
            },
            "toRef": {
                "id": "refs/heads/master",
                "repository": {
                    "slug": "my-repo",
                    "name": null,
                    "project": {
                        "key": "PRJ"
                    }
                }
            },
            "locked": false,
            "author": {
                "user": {
                    "name": "tom",
                    "emailAddress": "tom@example.com",
                    "id": 115026,
                    "displayName": "Tom",
                    "active": true,
                    "slug": "tom",
                    "type": "NORMAL"
                },
                "role": "AUTHOR",
                "approved": true,
                "status": "APPROVED"
            },
            "reviewers": [
                {
                    "user": {
                        "name": "jcitizen",
                        "emailAddress": "jane@example.com",
                        "id": 101,
                        "displayName": "Jane Citizen",
                        "active": true,
                        "slug": "jcitizen",
                        "type": "NORMAL"
                    },
                    "lastReviewedCommit": "7549846524f8aed2bd1c0249993ae1bf9d3c9998",
                    "role": "REVIEWER",
                    "approved": true,
                    "status": "APPROVED"
                }
            ],
            "participants": [
                {
                    "user": {
                        "name": "harry",
                        "emailAddress": "harry@example.com",
                        "id": 99049120,
                        "displayName": "Harry",
                        "active": true,
                        "slug": "harry",
                        "type": "NORMAL"
                    },
                    "role": "PARTICIPANT",
                    "approved": true,
                    "status": "APPROVED"
                },
                {
                    "user": {
                        "name": "dick",
                        "emailAddress": "dick@example.com",
                        "id": 3083181,
                        "displayName": "Dick",
                        "active": true,
                        "slug": "dick",
                        "type": "NORMAL"
                    },
                    "role": "PARTICIPANT",
                    "approved": false,
                    "status": "UNAPPROVED"
                }
            ],
            "links": {
                "self": [
                    {
                        "href": "http://link/to/pullrequest"
                    }
                ]
            }
        }
    ],
    "start": 0
}
```

</details>
If you prefer to avoid logging the email, I can log the same msg as 4 lines above i.e. the query params.

Reply:
Done, now the payload is logged via `log.Errorf` here and in all the other places. Not sure if you had another log level in mind (debug?)